substrate-node/runtime/src/migrations/tfgrid_v15_smart_contract_v8.rs

```diff
@@ -0,0 +1,231 @@
+use crate::pallet_smart_contract::{BalanceOf, Config, GRID_LOCK_ID};
+use crate::pallet_tfgrid::{AccountIdOf};
+use crate::sp_api_hidden_includes_construct_runtime::hidden_include::traits::{
+    Currency, LockableCurrency,
+};
+use crate::*;
+use frame_support::{
+    Blake2_128Concat,
+    pallet_prelude::OptionQuery,
+    storage_alias,
+    traits::OnRuntimeUpgrade,
+    traits::{Get, WithdrawReasons},
+    weights::Weight,
+};
+use log::{debug, info};
+use sp_std::{collections::btree_map::BTreeMap, marker::PhantomData};
+
+#[cfg(feature = "try-runtime")]
+use codec::Decode;
+#[cfg(feature = "try-runtime")]
+use sp_std::vec::Vec;
+
+// Storage alias from Twin v12
+#[storage_alias]
+pub type Twins<T: Config> = StorageMap<
+    pallet_tfgrid::Pallet<T>,
+    Blake2_128Concat,
+    pallet_tfgrid::TwinIndex,
+    pallet_tfgrid::migrations::types::v14::Twin<Vec<u8>, AccountIdOf<T>>,
+    OptionQuery,
+>;
+
+pub struct Migrate<T: Config>(PhantomData<T>);
+
+impl<T: Config> OnRuntimeUpgrade for Migrate<T> {
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<Vec<u8>, &'static str> {
+        info!(
+            "current pallet version: {:?}",
+            pallet_tfgrid::PalletVersion::<T>::get()
+        );
+        assert!(
+            pallet_tfgrid::PalletVersion::<T>::get()
+                >= pallet_tfgrid::types::StorageVersion::V14Struct
+        );
+
+        let twins_count: u64 = pallet_tfgrid::Twins::<T>::iter().count() as u64;
+        log::info!(
+            "🔎 MigrateTwinsV15 pre migration: Number of existing twins {:?}",
+            twins_count
+        );
+
+        info!("👥  TFGrid pallet to v14 passes PRE migrate checks ✅",);
+        Ok(twins_count.encode())
+    }
+
+    fn on_runtime_upgrade() -> Weight {
+        let mut twins: BTreeMap<u32, pallet_tfgrid::types::Twin<AccountIdOf<T>>> = BTreeMap::new();
+        let mut total_reads = 0;
+        let mut total_writes = 0;
+        if pallet_tfgrid::PalletVersion::<T>::get()
+            == pallet_tfgrid::types::StorageVersion::V14Struct
+        {
+            let (reads, writes) = migrate_tfgrid::<T>(&mut twins);
+            total_reads += reads;
+            total_writes += writes;
+        } else {
+            info!(" >>> Unused TFGrid pallet V15 migration");
+        }
+        if pallet_smart_contract::PalletVersion::<T>::get()
```

Comment:
I guess `pallet_tfgrid::PalletVersion::<T>::get() == pallet_tfgrid::types::StorageVersion::V14Struct` is also required here, so combine the if?